enterprise/coderd/userauth_test.go

```diff
@@ -31,128 +28,124 @@ func TestUserOIDC(t *testing.T) {
 	t.Run("RoleSync", func(t *testing.T) {
 		t.Parallel()
 
+		// NoRoles is the "control group". It has claims with 0 roles
+		// assigned, and asserts that the user has no roles.
 		t.Run("NoRoles", func(t *testing.T) {
 			t.Parallel()
 
-			ctx := testutil.Context(t, testutil.WaitMedium)
-			conf := coderdtest.NewOIDCConfig(t, "")
-
-			oidcRoleName := "TemplateAuthor"
-
-			config := conf.OIDCConfig(t, jwt.MapClaims{}, func(cfg *coderd.OIDCConfig) {
-				cfg.UserRoleMapping = map[string][]string{oidcRoleName: {rbac.RoleTemplateAdmin(), rbac.RoleUserAdmin()}}
-			})
-			config.AllowSignups = true
-			config.UserRoleField = "roles"
-
-			client, _ := coderdenttest.New(t, &coderdenttest.Options{
-				Options: &coderdtest.Options{
-					OIDCConfig: config,
-				},
-				LicenseOptions: &coderdenttest.LicenseOptions{
-					Features: license.Features{codersdk.FeatureUserRoleManagement: 1},
+			runner := setupOIDCTest(t, oidcTestConfig{
+				Config: func(cfg *coderd.OIDCConfig) {
+					cfg.AllowSignups = true
+					cfg.UserRoleField = "roles"
 				},
 			})
 
-			admin, err := client.User(ctx, "me")
-			require.NoError(t, err)
-			require.Len(t, admin.OrganizationIDs, 1)
-
-			resp := oidcCallback(t, client, conf.EncodeClaims(t, jwt.MapClaims{
+			claims := jwt.MapClaims{
 				"email": "alice@coder.com",
-			}))
-			require.Equal(t, http.StatusTemporaryRedirect, resp.StatusCode)
-			user, err := client.User(ctx, "alice")
-			require.NoError(t, err)
-
-			require.Len(t, user.Roles, 0)
-			roleNames := []string{}
-			require.ElementsMatch(t, roleNames, []string{})
+			}
+			// Login a new client that signs up
+			client, resp := runner.Login(t, claims)
+			require.Equal(t, http.StatusOK, resp.StatusCode)
+			// User should be in 0 groups.
+			runner.AssertRoles(t, "alice", []string{})
+			// Force a refresh, and assert nothing has changes
+			runner.ForceRefresh(t, client, claims)
+			runner.AssertRoles(t, "alice", []string{})
 		})
 
-		t.Run("NewUserAndRemoveRoles", func(t *testing.T) {
+		// A user has some roles, then on an oauth refresh will lose said
+		// roles from an updated claim.
+		t.Run("NewUserAndRemoveRolesOnRefresh", func(t *testing.T) {
+			// TODO: Implement new feature to update roles/groups on OIDC
+			// refresh tokens.
```

Comment:
follow-up: link issue for this

Reply:
https://github.com/coder/coder/issues/9312

Will add to comment